### PR TITLE
Feat/toplyne destination action

### PR DESCRIPTION
**Pull request by ankitjena**
The pull request is to add a new destination action for [Toplyne](https://toplyne.io). Using this segment users can directly send their data(from track, identify and group calls) to Toplyne. This will help Toplyne and it's users receive realtime data from segment without going through third party tools.

## Testing

We have tested end to end working of the actions included which go through our server and end up in our data warehouse.

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [ ] [Segmenters] Tested in the staging environment

### CODE REVIEW

**Comment by joe-ayoub-segment** [comment]
Hi @ankitjena , 
Thank you for raising this PR for a new Integration! 
We've updated our process slightly to include a face to face meeting so that the reviewer can get a better idea of the intended behavior and use-cases for the new Integration. This will help us do a more informed review. 

Would you be OK scheduling a call with me via [Calendly](http://calendly.com/joe_ayoub/integration-workshop) please? 
If you can't find a time that works for you please let me know (and provide some suggestions) and I'll open up some more slots in the calendar. 

Thank you!

File: packages/destination-actions/src/destinations/toplyne/sendUserProfiles/index.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+      required: true,
+      default: {
+        '@if': {
+          exists: { '@path': '$.userId' },
+          then: { '@path': '$.userId' },
+          else: { '@path': '$.anonymousId' }
+        }
+      }
+    },
+    creationTime: {
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/toplyne/sendUserProfiles/index.ts:26]: Please update the description to explain how creationTime works (as discussed)....

File: packages/destination-actions/src/destinations/toplyne/sendUserProfiles/index.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  title: 'Send user profiles',
+  description: 'Send identify calls to Toplyne',
+  defaultSubscription: 'type = "identify"',
+  fields: {
+    userId: {
+      type: 'string',
+      label: 'User ID',
+      description: 'The ID of the user to send properties for',
+      required: true,
+      default: {
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/toplyne/sendUserProfiles/index.ts:18]: How will we manage the anonymous => known user journey?

File: packages/destination-actions/src/destinations/toplyne/sendAccountProfiles/index.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import chunk from 'lodash/chunk'
+import { baseUrl } from '../constants'
+import type { Settings } from '../generated-types'
+import type { Payload } from './generated-types'
+
+const action: ActionDefinition<Settings, Payload> = {
+  title: 'Send Account Profiles',
+  description: 'Send group calls to Toplyne',
+  defaultSubscription: 'type = "group"',
+  fields: {
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/toplyne/sendAccountProfiles/index.ts:12]: Consider collecting the user identifier (userId or anonymousId) in order to link the account object to the user object, to compensate for lack of context.groupId in track() events.

**Review by joe-ayoub-segment — CHANGES_REQUESTED** [review]
Generally looks very good. 

Most important item to address is the identifiers. See inline comment. 

Consider if you want to collect context object data

**Comment by ankitjena** [comment]
@joe-ayoub-segment is there some documentation on how to add presets?

**Comment by joe-ayoub-segment** [comment]
Hi @ankitjena , 
Unfortunately there isn't. I'll be updating our docs in the next few weeks but this will be too late for you. 
Also, you can't test presets locally. 

However the following code should work. It needs to be added into the Destination object in your root index.ts file. 

`  presets: [
    {
      name: 'Send user profiles',
      subscribe: 'type = "identify"',
      partnerAction: 'sendUserProfiles',
      mapping: defaultValues(sendUserProfiles.fields)
    },
    {
      name: 'Send Account Profiles',
      subscribe: 'type = "group"',
      partnerAction: 'sendAccountProfiles',
      mapping: defaultValues(sendAccountProfiles.fields)
    },
    {
      name: 'Send events',
      subscribe: 'type = "track"',
      partnerAction: 'sendEvents',
      mapping: defaultValues(sendEvents.fields)
    }
  ]`

Best regards,
Joe

**Comment by joe-ayoub-segment** [comment]
hi @ankitjena , 
Hope you are doing well. Just following up to see if you need any assistance with this?
Thanks,
Joe

**Comment by ankitjena** [comment]
@joe-ayoub-segment Sorry for the delay, we were changing some things internally to support anonymousId thing. Have updated the PR. Can you re review?

Optimistic if we can be in the release tomorrow as well

File: packages/destination-actions/src/destinations/toplyne/sendAccountProfiles/index.ts (hunk 3 above)
Comment (ankitjena) [reply packages/destination-actions/src/destinations/toplyne/sendAccountProfiles/index.ts:12]: Not doing this now, we have a way of grouping information internally. We will change the process a bit and will update this destination accordingly.

File: packages/destination-actions/src/destinations/toplyne/sendUserProfiles/index.ts (hunk 2 above)
Comment (ankitjena) [reply packages/destination-actions/src/destinations/toplyne/sendUserProfiles/index.ts:18]: Done

File: packages/destination-actions/src/destinations/toplyne/sendUserProfiles/index.ts (hunk 1 above)
Comment (ankitjena) [reply packages/destination-actions/src/destinations/toplyne/sendUserProfiles/index.ts:26]: Done

File: packages/destination-actions/src/destinations/toplyne/index.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+import sendUserProfiles from './sendUserProfiles'
+
+import sendAccountProfiles from './sendAccountProfiles'
+
+import { baseUrl } from './constants'
+import sendEvents from './sendEvents'
+
+const destination: DestinationDefinition<Settings> = {
+  name: 'Toplyne',
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/toplyne/index.ts:12]:
```suggestion
  name: 'Toplyne Cloud Mode (Actions)',
```

File: packages/destination-actions/src/destinations/toplyne/index.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  slug: 'toplyne',
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/toplyne/index.ts:13]:
```suggestion
  slug: 'actions-toplyne-cloud',
```

**Comment by joe-ayoub-segment** [comment]
Hi @ankitjena I just took a look at the code and there hasn't been a recent commit, so nothing has changed. Are you sure that some changes have been submitted?

**Comment by joe-ayoub-segment** [comment]
Also @ankitjena  there's a couple of naming changes which will need to be added (which I missed the first time around). You can see them as comments in the root level index.ts file.

**Review by joe-ayoub-segment — CHANGES_REQUESTED** [review]

**Review by joe-ayoub-segment — CHANGES_REQUESTED** [review]
Hi @ankitjena , 
It all looks good apart from one thing. I discussed the performBatch() with a colleague of mine, and chunking and then sending multiple requests from the performBatch will mess with the deliverability observability for the Integration. 

I'm following up to see if we can set specific batch size limits for different Destinations from our side, so in that case you would not need to chunk. 

However for now we'll need to remove the performBatch(). 

Is this something you could do please?
Remember to also remove the enableBatch field as well. 

Everything else looks good. 
Thank you!

**Comment by ankitjena** [comment]
@joe-ayoub-segment That makes sense. Have removed it for now. Do verify once.

Also is there a way for us to know once you sort out the different batch sizes for destinations. Would definitely want to add it back then.

**Review by joe-ayoub-segment — APPROVED** [review]
Local tests pass. 
code looks good. 
New Integration so low risk.

**Comment by joe-ayoub-segment** [comment]
Hi @ankitjena  - PR merged. We should be able to deploy later today!
I'm not aware that we have a mechanism for keeping Partner informed about specific feature requests. I can add a note to the JIRA ticket to ping you once it's live though. 

At the moment it is just a request and hasn't been committed to - but improving batch handling and error reporting is a focus for the team at the moment.

**Comment by ankitjena** [comment]
@joe-ayoub-segment Thanks for the quick turnaround. Sure, I understand how feature requests work, so will be on lookout if docs are updated or anything.

Also the documentation PR is still open, had made the changes requested. As I understand both of these need go together. Pasting the PR link here https://github.com/segmentio/segment-docs/pull/4294

Will be looking forward to the release.

**Comment by joe-ayoub-segment** [comment]
Thanks @ankitjena , 
Another colleague of mine will liaise with you regarding docs. 
Regarding the missed call earlier. Do you want to reschedule?

**Comment by joe-ayoub-segment** [comment]
FYI @ankitjena today's deployment has been postponed due to an ongoing Datadog incident. 
We'll attempt to deploy tomorrow.

**Comment by ankitjena** [comment]
@joe-ayoub-segment Yeah, we were also facing issue in monitoring due to that. And apologies I forgot to cancel our call. For now don't think we need to reschedule.

Also I had discussion with our team, we can enable batch support by making changes to our apis. So it doesn't have to chunk internally. As I understand segment's batch size is 1000 right.

**Comment by joe-ayoub-segment** [comment]
Hi Ankit,
That's correct, Segment MAX MAX MAX batch size is 1000. However batch sizes
will usually be far smaller in reality.
Best regards,
Joe

**Comment by joe-ayoub-segment** [comment]
Hi @ankitjena FYI we didn't get to deploy yesterday - deployment pushed to Monday.

**Comment by joe-ayoub-segment** [comment]
hi @ankitjena this PR has been deployed. 
next steps are to do testing, create documentation (if you haven't already) and get the Catalog metadata created (if you haven't already). 

I'll send you an email with some details. 
Best regards,
Joe

**Pull request merged**